extractor/extractor.go

```diff
@@ -1658,3 +1768,91 @@ func extractNumLines(tw *trap.Writer, fileName string, ast *ast.File) {
 
 	dbscheme.NumlinesTable.Emit(tw, tw.Labeler.FileLabel(), lineCount, linesOfCode, linesOfComments)
 }
+
+// For a type `t` which is the type of a field of an interface type, return
+// whether `t` a type set literal which is not a union type. Note that a field
+// of an interface must be a method signature, an embedded interface type or a
+// type set literal.
+func isNonUnionTypeSetLiteral(t types.Type) bool {
+	switch t.Underlying().(type) {
+	case *types.Interface, *types.Union, *types.Signature:
+		return false
+	default:
+		return true
+	}
+}
+
+// Given a type `t`, return a union with a single term that is `t` without a
+// tilde.
+func createUnionFromType(t types.Type) *types.Union {
+	return types.NewUnion([]*types.Term{types.NewTerm(false, t)})
+}
+
+// Go through a `FieldList` and update the types of all type set literals which
+// are not already union types to be union types. We do this by changing the
+// types stored in `tw.Package.TypesInfo.Types`.
+func overrideTypesOfTypeSetLiterals(tw *trap.Writer, fields *ast.FieldList) {
```

Comment:
```suggestion
func makeTypeSetLiteralsUnionTyped(tw *trap.Writer, fields *ast.FieldList) {
```

Probably add a comment at the use sites explaining the reasoning why we're confident we don't extract a type-set literal prior to this override being applied.